core/src/main/java/io/questdb/cairo/mv/SampleByRangeCursor.java

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.mv;
+
+import io.questdb.griffin.engine.groupby.TimestampSampler;
+import io.questdb.std.Numbers;
+import io.questdb.std.datetime.TimeZoneRules;
+import io.questdb.std.datetime.microtime.Timestamps;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Allows iterating through SAMPLE BY buckets with the given step.
+ * The goal is to split a potentially large time interval to be scanned by
+ * the materialized view query into smaller intervals, thus minimizing
+ */
+public class SampleByRangeCursor {
+    private long maxTimestamp;
+    private long minTimestamp;
+    private TimestampSampler sampler;
+    private int step;
+    private long timestampHi;
+    private long timestampLo;
+    private TimeZoneRules tzRules;
+
+    public long getMaxTimestamp() {
+        return maxTimestamp;
+    }
+
+    public long getMinTimestamp() {
+        return minTimestamp;
+    }
+
+    public long getTimestampHi() {
+        return timestampHi;
+    }
+
+    public long getTimestampLo() {
+        return timestampLo;
+    }
+
+    public boolean next() {
+        if (timestampHi != maxTimestamp) {
+            timestampLo = timestampHi;
+            for (int i = 0; i < step; i++) {
+                timestampHi = nextTimestamp(timestampHi);
+                if (timestampHi == maxTimestamp) {
```

Comment:
this loop exits when i = step (not a good variable name) or timestamp reaches timestampHi - which is immovable target, i don't see how this look exists "step" count of times

Reply:
Not sure if I understand the question. What we do here is the following: we try to move the hi range boundary `step` number of times and that's our iterator's single step (hence, the variable name). If we reach the max boundary on the way, we stop and end up with the `[timestampLo, maxTimestamp]` interval.